File: src/test/java/fr/adrienbrault/idea/symfonyplugin/tests/dic/inspection/MissingServiceInspectionTest.java
package fr.adrienbrault.idea.symfonyplugin.tests.dic.inspection;

import fr.adrienbrault.idea.symfonyplugin.dic.inspection.MissingServiceInspection;
import fr.adrienbrault.idea.symfonyplugin.tests.SymfonyLightCodeInsightFixtureTestCase;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see fr.adrienbrault.idea.symfonyplugin.dic.inspection.MissingServiceInspection
 */
public class MissingServiceInspectionTest extends SymfonyLightCodeInsightFixtureTestCase {
    public void setUp() throws Exception {
        super.setUp();
        myFixture.copyFileToProject("classes.php");
        myFixture.copyFileToProject("services.xml");
    }

    public String getTestDataPath() {
        return "src/test/java/fr/adrienbrault/idea/symfonyplugin/tests/dic/inspection/fixtures";
    }

    public void testThatPhpServiceInterfaceForGetMethodIsInspected() {
        assertLocalInspectionContains("test.php", "<?php\n" +
                "/** @var $x \\Symfony\\Component\\DependencyInjection\\ContainerInterface */\n" +
                "$x->get('fo<caret>obar')",
            MissingServiceInspection.INSPECTION_MESSAGE
        );

        assertLocalInspectionNotContains("test.php", "<?php\n" +
                "/** @var $x \\Symfony\\Component\\DependencyInjection\\ContainerInterface */\n" +
                "$x->get('app.ma<caret>iler')",
            MissingServiceInspection.INSPECTION_MESSAGE
        );
    }

    public void testThatYamlServiceInterfaceForGetMethodIsInspected() {
        assertLocalInspectionContains("services.yml", "services:\n   @args<caret>_unknown", "Missing Service");
        assertLocalInspectionContains("services.yml", "services:\n   @Args<caret>_unknown", "Missing Service");

        assertLocalInspectionNotContains("services.yml", "services:\n   @App.ma<caret>iler", "Missing Service");
        assertLocalInspectionNotContains("services.yml", "services:\n   @app.ma<caret>iler", "Missing Service");

        assertLocalInspectionNotContains("services.yml", "services:\n   @@args<caret>_unknown", "Missing Service");
        assertLocalInspectionNotContains("services.yml", "services:\n   @=args<caret>_unknown", "Missing Service");
    }
}

File: src/test/kotlin/fr/adrienbrault/idea/symfony2plugin/tests/dic/inspection/MissingServiceInspectionTest.kt
package fr.adrienbrault.idea.symfony2plugin.tests.dic.inspection

import fr.adrienbrault.idea.symfony2plugin.dic.inspection.MissingServiceInspection
import fr.adrienbrault.idea.symfony2plugin.tests.SymfonyLightCodeInsightFixtureTestCase

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see fr.adrienbrault.idea.symfony2plugin.dic.inspection.MissingServiceInspection
 */
class MissingServiceInspectionTest : SymfonyLightCodeInsightFixtureTestCase() {
    override fun setUp() {
        super.setUp()
        myFixture.copyFileToProject("classes.php")
        myFixture.copyFileToProject("services.xml")
    }

    override fun getTestDataPath(): String {
        return "src/test/kotlin/fr/adrienbrault/idea/symfony2plugin/tests/dic/inspection/fixtures"
    }

    fun testThatPhpServiceInterfaceForGetMethodIsInspected() {
        assertLocalInspectionContains(MissingServiceInspection.PhpLocalInspectionTool::class.java, "test.php", "<?php\n" +
                "/** @var \$x \\Symfony\\Component\\DependencyInjection\\ContainerInterface */\n" +
                "\$x->get('fo<caret>obar')",
            MissingServiceInspection.INSPECTION_MESSAGE
        )

        assertLocalInspectionNotContains(MissingServiceInspection.PhpLocalInspectionTool::class.java, "test.php", "<?php\n" +
                "/** @var \$x \\Symfony\\Component\\DependencyInjection\\ContainerInterface */\n" +
                "\$x->get('app.ma<caret>iler')",
            MissingServiceInspection.INSPECTION_MESSAGE
        )
    }

    fun testThatContainerBagGetMethodIsNotInspectedAsService() {
        assertLocalInspectionNotContains(MissingServiceInspection.PhpLocalInspectionTool::class.java, "test.php", "<?php\n" +
                "use Symfony\\Component\\DependencyInjection\\ParameterBag\\ContainerBagInterface;\n" +
                "\n" +
                "class TestService\n" +
                "{\n" +
                "    public function __construct(private ContainerBagInterface \$containerBag) {}\n" +
                "\n" +
                "    public function testFoo()\n" +
                "    {\n" +
                "        \$this->containerBag->get('app.to<caret>ken');\n" +
                "    }\n" +
                "}",
            MissingServiceInspection.INSPECTION_MESSAGE
        )
    }

    fun testThatPhpAttributesForServiceAutowireIsInspected() {
        assertLocalInspectionContains(MissingServiceInspection.PhpLocalInspectionTool::class.java, "test.php", "<?php\n" +
                "use Symfony\\Component\\DependencyInjection\\Attribute\\Autowire;\n" +
                "\n" +
                "class HandlerCollection\n" +
                "{\n" +
                "    public function __construct(\n" +
                "        #[Autowire(service: 'fo<caret>obar')]" +
                "    ) {}\n" +
                "}",
            MissingServiceInspection.INSPECTION_MESSAGE
        )

        assertLocalInspectionContains(MissingServiceInspection.PhpLocalInspectionTool::class.java, "test.php", "<?php\n" +
                "use Symfony\\Component\\DependencyInjection\\Attribute\\Autowire;\n" +
                "\n" +
                "class HandlerCollection\n" +
                "{\n" +
                "    public function __construct(\n" +
                "        #[Autowire(service: \"fo<caret>obar\")]" +
                "    ) {}\n" +
                "}",
            MissingServiceInspection.INSPECTION_MESSAGE
        )
    }

    fun testThatPhpAttributesForServiceAsDecoratorIsInspected() {
        assertLocalInspectionContains(MissingServiceInspection.PhpLocalInspectionTool::class.java, "test.php", "<?php\n" +
                "use Symfony\\Component\\DependencyInjection\\Attribute\\AsDecorator;\n" +
                "#[AsDecorator(\"fo<caret>obar\")]\n" +
                "class HandlerCollection {}",
            MissingServiceInspection.INSPECTION_MESSAGE
        )
    }

    fun testThatYamlServiceInterfaceForGetMethodIsInspected() {
        assertLocalInspectionContains(MissingServiceInspection.YamlLocalInspectionTool::class.java, "services.yml", "services:\n   @args<caret>_unknown", MissingServiceInspection.INSPECTION_MESSAGE)
        assertLocalInspectionContains(MissingServiceInspection.YamlLocalInspectionTool::class.java, "services.yml", "services:\n   @Args<caret>_unknown", MissingServiceInspection.INSPECTION_MESSAGE)

        assertLocalInspectionNotContains(MissingServiceInspection.YamlLocalInspectionTool::class.java, "services.yml", "services:\n   @App.ma<caret>iler", MissingServiceInspection.INSPECTION_MESSAGE)
        assertLocalInspectionNotContains(MissingServiceInspection.YamlLocalInspectionTool::class.java, "services.yml", "services:\n   @app.ma<caret>iler", MissingServiceInspection.INSPECTION_MESSAGE)

        assertLocalInspectionNotContains(MissingServiceInspection.YamlLocalInspectionTool::class.java, "services.yml", "services:\n   @@args<caret>_unknown", MissingServiceInspection.INSPECTION_MESSAGE)
        assertLocalInspectionNotContains(MissingServiceInspection.YamlLocalInspectionTool::class.java, "services.yml", "services:\n   @=args<caret>_unknown", MissingServiceInspection.INSPECTION_MESSAGE)
    }
}
